### PR TITLE
[#145612229] add session state caching support for geode

### PR DESCRIPTION
**Pull request by pivotal-Jammy-Louie**
Pull request to Support Session State Caching for Pivotal Cloud Cache (Gemfire)

### CODE REVIEW

**Comment by pivotal-Jammy-Louie** [comment]
Updated the branch not to ignore .idea/ and include the infrastructural files

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+module JavaBuildpack
+  module Container
+
+    # Encapsulates the detect, compile, and release functionality for Tomcat Redis support.
+    class TomcatGeodeStore < JavaBuildpack::Component::VersionedDependencyComponent
+      include JavaBuildpack::Container
+
+      # (see JavaBuildpack::Component::VersionedDependencyComponent#supports?)
+      def supports?
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: Method should be [protected](https://github.com/cloudfoundry/java-buildpack/blob/master/lib/java_buildpack/framework/new_relic_agent.rb#L46-L53)

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        @application.services.one_service? FILTER, KEY_LOCATORS, KEY_USERS
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#release)
+      def release
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: sorted alphabetically within visibility, e.g. `release` after `compile`.

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+        @droplet.java_opts.add_system_property 'gemfire.security-username', 'cluster_operator'
+        @droplet.java_opts.add_system_property 'gemfire.security-password', user['password']
+        @droplet.java_opts.add_system_property 'gemfire.security-client-auth-init',
+                                               'io.pivotal.cloudcache.ClientAuthInitialize.create'
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#compile)
+      def compile
+        return unless supports?
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: No need for this (that's why you don't see it anywhere else).  `compile` isn't called unless `detect` first confirms its use.

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+      include JavaBuildpack::Container
+
+      # (see JavaBuildpack::Component::VersionedDependencyComponent#supports?)
+      def supports?
+        @application.services.one_service? FILTER, KEY_LOCATORS, KEY_USERS
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#release)
+      def release
+        return unless supports?
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: No need for this (that's why you don't see it anywhere else).  `release` isn't called unless `detect` first confirms its use.

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      SCHEMA_INSTANCE_URL = 'http://www.w3.org/2001/XMLSchema-instance'.freeze
+      SCHEMA_LOCATION = 'http://geode.apache.org/schema/cache http://geode.apache.org/schema/cache/cache-1.0.xsd'.freeze
+      LOCATOR_REGEXP = Regexp.new("([^\\[]+)\\[([^\\]]+)\\]").freeze
+      FUNCTION_SERVICE_CLASS_NAMES = [
+        'org.apache.geode.modules.util.CreateRegionFunction',
+        'org.apache.geode.modules.util.TouchPartitionedRegionEntriesFunction',
+        'org.apache.geode.modules.util.TouchReplicatedRegionEntriesFunction',
+        'org.apache.geode.modules.util.RegionSizeFunction'].freeze
+
+      private_constant :FILTER, :KEY_LOCATORS, :KEY_USERS
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: All constants above need to be `private_constant`.

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+      def cache_client_xml_path
+        @droplet.sandbox + 'conf' + cache_client_xml
+      end
+
+      def cache_client_xml
+        'cache-client.xml'
+      end
+
+      FILTER = /session-replication/
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: Private constants all go before private methods.

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      LOCATOR_REGEXP = Regexp.new("([^\\[]+)\\[([^\\]]+)\\]").freeze
+      FUNCTION_SERVICE_CLASS_NAMES = [
+        'org.apache.geode.modules.util.CreateRegionFunction',
+        'org.apache.geode.modules.util.TouchPartitionedRegionEntriesFunction',
+        'org.apache.geode.modules.util.TouchReplicatedRegionEntriesFunction',
+        'org.apache.geode.modules.util.RegionSizeFunction'].freeze
+
+      private_constant :FILTER, :KEY_LOCATORS, :KEY_USERS
+
+      def tar_name
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: Methods alphabetized with visibilities

File: lib/java_buildpack/container/tomcat/tomcat_instance.rb [hunk 8]
```diff
@@ -93,7 +93,6 @@ def expand(file)
         with_timing "Expanding #{@component_name} to #{@droplet.sandbox.relative_path_from(@droplet.root)}" do
           FileUtils.mkdir_p @droplet.sandbox
           shell "tar xzf #{file.path} -C #{@droplet.sandbox} --strip 1 --exclude webapps 2>&1"
-
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_instance.rb]: This line should not be removed.

File: lib/java_buildpack/container/tomcat/tomcat_redis_store.rb [hunk 9]
```diff
@@ -107,8 +107,6 @@ def mutate_context
 
         write_xml context_xml, document
       end
-
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_redis_store.rb]: This line should not be removed.

File: lib/java_buildpack/container/tomcat/tomcat_redis_store.rb [hunk 10]
```diff
@@ -107,8 +107,6 @@ def mutate_context
...
     end
-
```
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_redis_store.rb]: This line should not be removed.

File: spec/java_buildpack/container/tomcat/tomcat_geode_store_spec.rb [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    component.compile
+
+    expect(sandbox + "lib/stub-geode-store/stub-jar-1.jar").not_to exist
+    expect(sandbox + "lib/stub-geode-store/stub-jar-2.jar").not_to exist
+  end
+
+  context 'when there is a session-replication service' do
+    before do
+      allow(services).to receive(:one_service?).with(/session-replication/, 'locators', 'users')
+        .and_return(true)
```
Comment (nebhale) [inline spec/java_buildpack/container/tomcat/tomcat_geode_store_spec.rb]: You'll want to run `bundle exec rake` to ensure that your code meets all formatting requirements.  The Rubocop task will give you pointers on where the code needs to change.

File: spec/java_buildpack/container/tomcat_spec.rb [hunk 12]
```diff
@@ -19,6 +19,7 @@
 require 'java_buildpack/container/tomcat'
 require 'java_buildpack/container/tomcat/tomcat_insight_support'
 require 'java_buildpack/container/tomcat/tomcat_instance'
+require 'java_buildpack/container/tomcat/tomcat_geode_store'
```
Comment (nebhale) [inline spec/java_buildpack/container/tomcat_spec.rb]: Alphabetize

File: spec/java_buildpack/container/tomcat_spec.rb [hunk 13]
```diff
@@ -32,6 +33,7 @@
   let(:configuration) do
     { 'tomcat'                 => tomcat_configuration,
       'lifecycle_support'      => lifecycle_support_configuration,
+      'geode_store'            => geode_store_configuration,
```
Comment (nebhale) [inline spec/java_buildpack/container/tomcat_spec.rb]: Alphabetize

File: spec/java_buildpack/container/tomcat_spec.rb [hunk 14]
```diff
@@ -73,6 +77,8 @@
       .to receive(:new).with(sub_configuration_context(tomcat_configuration))
     allow(JavaBuildpack::Container::TomcatLifecycleSupport)
       .to receive(:new).with(sub_configuration_context(lifecycle_support_configuration))
+    allow(JavaBuildpack::Container::TomcatGeodeStore)
```
Comment (nebhale) [inline spec/java_buildpack/container/tomcat_spec.rb]: Alphabetize

File: spec/java_buildpack/container/tomcat_spec.rb [hunk 15]
```diff
@@ -44,6 +46,8 @@
 
   let(:logging_support_configuration) { instance_double('logging-support-configuration') }
 
+  let(:geode_store_configuration) { instance_double('geode_store_configuration') }
```
Comment (nebhale) [inline spec/java_buildpack/container/tomcat_spec.rb]: Alphabetize

File: config/tomcat.yml [hunk 16]
```diff
@@ -39,3 +39,6 @@ redis_store:
   database: 0
   timeout: 2000
   connection_pool_size: 2
+geode_store:
+  version: 0.+
+  repository_root: "https://tomcat-javabuildpack-spike.s3.amazonaws.com/geode-store"
```
Comment (pivotal-Jammy-Louie) [inline config/tomcat.yml]: in the process of find out where i can place the tar

Comment (nebhale) [inline config/tomcat.yml]: Sounds good.

Comment (nebhale) [inline config/tomcat.yml]: If the code really is Geode, and is all backed by an ASL repository, we could host it for you.

**Comment by nebhale** [comment]
@pivotal-Jammy-Louie Also noticed that you need to take care of [publicizing your CLA compliance](https://github.com/cloudfoundry/java-buildpack/pull/437#issuecomment-305807423).

**Comment by nebhale** [comment]
OK, looks like just the repository outstanding and then we can merge.  Let's chat Monday morning about what the options are there.

File: config/tomcat.yml (hunk 16 above)
Comment (pivotal-Jammy-Louie) [inline config/tomcat.yml]: all but one jar is from Geode

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb (hunk 4 above)
Comment (pivotal-Jammy-Louie) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: when I tried to remove the check for supports i get the following

 ERROR Release failed with exception #<NoMethodError: undefined method `[]' for nil:NilClass>

File: config/tomcat.yml (hunk 16 above)
Comment (nebhale) [inline config/tomcat.yml]: Where is that one file from?  Is it GemFire branded?

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb (hunk 4 above)
Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: Under what circumstances, tests?  Or in actual usage?

File: config/tomcat.yml (hunk 16 above)
Comment (pivotal-Jammy-Louie) [inline config/tomcat.yml]: the jar is called pcc-client-auth

File: lib/java_buildpack/container/tomcat/tomcat_geode_store.rb (hunk 4 above)
Comment (pivotal-Jammy-Louie) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: it fails on actual usage when attempting to compile.

Comment (nebhale) [inline lib/java_buildpack/container/tomcat/tomcat_geode_store.rb]: Yep, you're right.  I forgot that 1`ModularComponents` behave differently with respect to their modules.  You can disregard these comments.

File: config/tomcat.yml (hunk 16 above)
Comment (nebhale) [inline config/tomcat.yml]: Yeah, it'll need to be a public repository with a permissive license for us to host it.  This goes back to the fact that the CFF needs to take a special exception for US export compliance that can only be satisfied by 100% open source code.

**Comment by nebhale** [comment]
@pivotal-Jammy-Louie Can you please publicize your org so that the CLA is satisfied?

**Comment by nebhale** [comment]
@pivotal-Jammy-Louie We're still waiting on you to publicize your GitHub org membership to identify that you've signed the CLA.

**Comment by nebhale** [comment]
Great.  The PR is more or less ready to go now, after we determine the hosting thing.

**Comment by pulkitpivotal** [comment]
I think this PR is ready to be merged, please let us know @nebhale if you have more feedback on it.

**Comment by nebhale** [comment]
https://java-buildpack.cloudfoundry.org/geode-store/index.yml
https://java-buildpack.cloudfoundry.org/geode-store/geode-store-0.0.2.tar.gz

**Comment by luisapace** [comment]
@pivotal-Jammy-Louie 
Hi Jammy, 
I've had your name and this thread from Neb Hale, I kindly ask you if you can help me or redirect this to the responsible people.

I am an IBM employee working in legal team and my goal is to analyze with some software tools the source code and the license of all open sources and their dependencies to check if they can be included in our products without legal issues (in .cc a colleague working with me). 

Currently I'm analyzing the java buildpack 3.19 code to verify if we can include it in our code without legal problems; to do that I've to analyze it and all its dependencies chain. 

Now, the java buildpack 3.19 calls as dependency - with the name Apache Geode Tomcat Session Store - the zip file at:https://java-buildpack.cloudfoundry.org/geode-store/geode-store-0.0.2.tar.gz, that in turn include some jar files that seems to be NOT open source, but commercial code, Pivotal proprietary: 
geode-core-9.0.3.jar 
geode-modules-9.0.3.jar 
geode-modules-tomcat8-9.0.3.jar 
pcc-client-auth-0.0.1.jar

Could you please confirm me if they are commercial code or opensource? 

If they are opensource could you please point me to the download location of their source. 

Feel free to ask us other details if you need. 

Thanks for your support.

**Pull request closed without merging**